src/main/java/io/odpf/depot/http/request/builder/HeaderBuilder.java

```diff
@@ -1,16 +1,92 @@
 package io.odpf.depot.http.request.builder;
 
+import io.odpf.depot.common.Template;
+import io.odpf.depot.config.HttpSinkConfig;
+import io.odpf.depot.exception.InvalidTemplateException;
+import io.odpf.depot.http.enums.HttpParameterSourceType;
+import io.odpf.depot.message.MessageContainer;
+import io.odpf.depot.message.OdpfMessageParser;
+import io.odpf.depot.message.OdpfMessageSchema;
+import io.odpf.depot.message.ParsedOdpfMessage;
+import lombok.Getter;
+
+import java.io.IOException;
+import java.util.HashMap;
 import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import static io.odpf.depot.http.enums.HttpParameterSourceType.KEY;
 
+@Getter
 public class HeaderBuilder {
 
-    private final Map<String, String> headerConfig;
+    private final Map<String, String> baseHeaders;
+    private final Properties headersTemplateProperty;
+    private final HttpParameterSourceType headersParameterSource;
+    private final String schemaProtoKeyClass;
+    private final String schemaProtoMessageClass;
 
-    public HeaderBuilder(Map<String, String> headerConfig) {
-        this.headerConfig = headerConfig;
+    public HeaderBuilder(HttpSinkConfig config) {
+        this.baseHeaders = config.getSinkHttpHeaders();
+        this.headersTemplateProperty = config.getSinkHttpHeadersTemplate();
+        this.headersParameterSource = config.getSinkHttpHeadersParameterSource();
+        this.schemaProtoKeyClass = config.getSinkConnectorSchemaProtoKeyClass();
+        this.schemaProtoMessageClass = config.getSinkConnectorSchemaProtoMessageClass();
     }
 
     public Map<String, String> build() {
+        return baseHeaders;
+    }
+
+    public Map<String, String> build(MessageContainer container, OdpfMessageParser odpfMessageParser) throws IOException {
+        if (headersTemplateProperty == null || headersTemplateProperty.isEmpty()) {
+            return build();
+        }
+
+        Map<Template, Template> headersTemplateMap = createHeadersTemplateMap();
```

Comment:
this map template object should not be on build(). it should be set as constructor